x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/AutoFollowCoordinator.java

```
@@ -0,0 +1,293 @@
+                    FollowIndexAction.Request followRequest =
+                        new FollowIndexAction.Request(leaderIndexNameWithClusterAliasPrefix, followIndexName,
+                            autoFollowPattern.getMaxBatchOperationCount(), autoFollowPattern.getMaxConcurrentReadBatches(),
+                            autoFollowPattern.getMaxOperationSizeInBytes(), autoFollowPattern.getMaxConcurrentWriteBatches(),
+                            autoFollowPattern.getMaxWriteBufferSize(), autoFollowPattern.getRetryTimeout(),
+                            autoFollowPattern.getIdleShardRetryDelay());
+
+                    // Execute if the create and follow api call succeeds:
+                    Runnable successHandler = () -> {
+                        LOGGER.info("Auto followed leader index [{}] as follow index [{}]", leaderIndexName, followIndexName);
+
+                        // This function updates the auto follow metadata in the cluster to record that the leader index has been followed:
+                        // (so that we do not try to follow it in subsequent auto follow runs)
+                        Function<ClusterState, ClusterState> function = recordLeaderIndexAsFollowFunction(clusterAlias, indexToFollow);
+                        // The coordinator always runs on the elected master node, so we can update cluster state here:
+                        updateAutoFollowMetadata(function, updateError -> {
+                            if (updateError != null) {
+                                LOGGER.error("Failed to mark leader index [" + leaderIndexName + "] as auto followed", updateError);
+                            } else {
+                                LOGGER.debug("Successfully marked leader index [{}] as auto followed", leaderIndexName);
+                            }
+                            if (countDown.countDown()) {
+                                finalise(updateError);
```

So I think that something like this means that we are losing errors. So we have:
 - a pattern `metricbeat-*`
 - let's say that matches two indices in the leader cluster `metricbeat-2018-08-27` and `metricbeat-2018-08-28`
 - we have an outer `CountDown` `AutoFollower#countDown` initialized to one for the single pattern
 - we have an inner `CountDown` `countDown` in `handleClusterAlias` initialized to two for the two indices
 - the first cluster state update task for the first index fails
 - we count down `handleClusterAlias#countDown` and it decrements to one; we do nothing
 - the second cluster state update task for the second index succeeds, or fails; either way, we would count down `handleClusterAlias#countDown` again and the exception from the first failure would be lost

---

Good point. I'll keep track of the errorHolder and suppress subsequent errors to the first error if multiple exceptions occur.